File: c/beer_song.c
#include "beer_song.h"

#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

void recite(uint8_t start_bottles, uint8_t take_down, char **s) {
  size_t i = 0;
  while (take_down-- > 0) {
    if (start_bottles == 0) {
      sprintf(s[i++],
              "No more bottles of beer on the wall, no more bottles of beer.");
      sprintf(
          s[i++],
          "Go to the store and buy some more, 99 bottles of beer on the wall.");
    } else if (start_bottles == 1) {
      sprintf(s[i++], "%d bottle of beer on the wall, %d bottle of beer.",
              start_bottles, start_bottles);
      sprintf(s[i++], "Take it down and pass "
                      "it around, no more bottles of beer on the wall.");
    } else if (start_bottles == 2) {
      sprintf(s[i++], "%d bottles of beer on the wall, %d bottles of beer.",
              start_bottles, start_bottles);
      sprintf(s[i++],
              "Take one down and "
              "pass it around, %d bottle of beer on the wall.",
              start_bottles - 1);
    } else {
      sprintf(s[i++], "%d bottles of beer on the wall, %d bottles of beer.",
              start_bottles, start_bottles);
      sprintf(s[i++],
              "Take one down and "
              "pass it around, %d bottles of beer on the wall.",
              start_bottles - 1);
    }
    if (start_bottles > 0) {
      strcpy(s[i++], "");
    }
    start_bottles--;
  }
}

File: dart/beer_song.dart
class BeerSong {
  String aux(int i) {
    if (i == 0) {
      return 'no more bottles';
    }
    if (i == 1) {
      return '1 bottle';
    }
    return '$i bottles';
  }

  List<String> recite(int bottles, int start) {
    List<String> verses = [];
    for (int i = 0; i < start; i++) {
      if (i > 0) {
        verses.add('');
      }
      if (bottles == 0) {
        verses.add(
            'No more bottles of beer on the wall, no more bottles of beer.');
        verses.add(
            'Go to the store and buy some more, 99 bottles of beer on the wall.');
      } else {
        verses.add(
            '${aux(bottles)} of beer on the wall, ${aux(bottles)} of beer.');
        bottles--;
        verses.add(
            'Take ${bottles == 0 ? "it" : "one"} down and pass it around, ${aux(bottles)} of beer on the wall.');
      }
    }

    return verses;
  }
}
